db/crud.go

```diff
@@ -541,21 +541,19 @@ func (db *DatabaseCollectionWithUser) Get1xRevAndChannels(ctx context.Context, d
 }
 
 // Returns an HTTP 403 error if the User is not allowed to access any of this revision's channels.
-func (col *DatabaseCollectionWithUser) authorizeDoc(doc *Document, revid string) error {
+func (col *DatabaseCollectionWithUser) authorizeDoc(ctx context.Context, doc *Document, revid string) error {
 	user := col.user
 	if doc == nil || user == nil {
 		return nil // A nil User means access control is disabled
 	}
-	if revid == "" {
-		revid = doc.CurrentRev
-	}
-	if rev := doc.History[revid]; rev != nil {
-		// Authenticate against specific revision:
-		return col.user.AuthorizeAnyCollectionChannel(col.ScopeName, col.Name, rev.Channels)
-	} else {
+
+	channelsForRev, ok := doc.channelsForRev(revid)
```

Comment:
It looks like the effect here is that an authorizeDoc call for a existing, non-leaf rev is going to be treated as a 403 (because channelsForRev will be empty).  Is this a change in behaviour (i.e. it previously would have returned a 404)?

Reply:
I see what you mean.  The original comment on CBG-3203 was "Further analysis is required to identify any spots we're still attempting to reference channel assignment for non-leaf nodes - if those exist, they should be updated to use the channels associated with the active revision."  

What's your opinion on whether we should do what you've got here (return a 403) when attempting to authorize a non-leaf rev, or if we should authenticate against the channels on the active rev?

Reply:
Ah, I see where you were going with this comment now. Because we've stripped the non-leaf channels, there's an implicit change in behaviour from this function, rather than an explicit one. Prior to removal of revtree channels, clients could request an older rev and still get correct access checks from it.

This would affect _all_ of the in-flight old rev requests we handle. I don't think 403 is accurate nor suitable, so this code does need changing.

I think the options here would be to either:
- Use active rev's channels as an optimistic check (assume doc channels haven't changed).
  - Worst case, if channels have been added: Clients can pull a revision they didn't have access to, assuming they knew the rev ID (changes wouldn't have sent them the entry, since it's sequence based rather than rev ID based)
  - If doc channels have been removed: Clients get a 403 for an in-flight rev they would've had access to. Doesn't seem like a big deal, given they just lost doc access anyway.
- Return 404 (are in-flight revs considered best-effort given caching/TTL?)
  - If clients still have doc access they'll get a subsequent changes/rev anyway, and the in-flight rev would've been replaced soon.
  - If clients don't have doc access any more, they won't get an older rev they would've previously.

I'm leaning towards returning 404 for non-leaf revs, as it seems safer (in terms of not sending anything a client shouldn't have), but it does effectively remove the concept of sending in-flight revisions.

Reply:
As discussed, will return 404 in this case (all uncached in-flight rev requests)

May need to differentiate between no channels and unset channels from revcache to do this.